File: frontend/src/types/jlens.ts
/**
 * J-space lens wire format — the UI mirror of backend/src/schemas/jlens.py.
 *
 * THE WIRE FORMAT IS NOT OURS TO DESIGN (BR-029, PADR IDL-45). These shapes
 * mirror Neuronpedia's lens stream exactly, so the readout panel renders a
 * miStudio stream and a Neuronpedia stream with no adaptation layer. Adding a
 * miStudio-shaped field here would silently break that property — the panel
 * would keep working against our server and fail against theirs.
 *
 *   meta  = { model, types, layers_by_type, top_n, prompt_len }
 *   token = { position, token, id, is_generated, results: slice[] }
 *   slice = { type, top_tokens[layer][k], top_probs[layer][k] }
 *
 * `top_tokens` entries are DECODED STRINGS, not ids — the backend enforces
 * this because ids type-check against a looser schema and render as
 * unreadable cells.
 */

/**
 * Lens types the stream can CARRY.
 *
 * DIFF is deliberately absent: it is a client-side rendering mode over two
 * transported slices, never a transported type. Adding it here would invite a
 * request for a type the server cannot emit.
 */
export type LensType = 'JACOBIAN_LENS' | 'LOGIT_LENS';

/** What the mode tabs offer. DIFF exists only at this layer. */
export type LensMode = LensType | 'DIFF';

/**
 * Which computations are meaningful at one layer.
 *
 * Layer kind is PER-LAYER state, not a model property: a hybrid model
 * interleaves convolutional and attention layers, so "freeze Q/K" is undefined
 * on some of them. Inapplicable is `null` (absent), never `false` — a `false`
 * gets averaged by a consumer and silently understates (BR-032).
 */
export interface LayerApplicability {
  layer: number;
  has_attention: boolean;
  frozen_qk_applicable: boolean | null;
  broadcast_metrics_applicable: boolean | null;
}

/**
 * One lens type's readout for one token position, across layers.
 *
 * `top_tokens[layerIdx][k]` is indexed by POSITION IN `meta.layers_by_type[type]`,
 * not by the model's absolute layer number. Indexing it with an absolute layer
 * number reads the wrong row wherever the two differ, and produces a plausible
 * grid rather than an error.
 */
export interface LensTypeSlice {
  type: LensType;
  top_tokens: string[][];
  top_probs: number[][];
}

export interface LensMetaMessage {
  kind: 'meta';
  model: string;
  types: LensType[];
  /** Absolute layer indices per lens type. DRIVES THE LAYER AXIS — never assume
   *  a count or a spacing; models here range from 16 to 26+ layers. */
  layers_by_type: Record<string, number[]>;
  top_n: number;
  prompt_len: number;
  layer_applicability?: LayerApplicability[] | null;
}

export interface LensTokenMessage {
  kind: 'token';
  position: number;
  token: string;
  id: number;
  is_generated: boolean;
  results: LensTypeSlice[];
}

export interface ReadoutRequest {
  model_id: string;
  prompt: string;
  types?: LensType[];
  layers?: number[] | null;
  top_n?: number;
  /** Required by the server when `types` includes JACOBIAN_LENS. The logit lens
   *  needs no artifact (BR-005). */
  artifact_id?: string | null;
}

/** Non-streaming envelope. CONTAINS a meta message rather than being one. */
export interface ReadoutResponse {
  meta: LensMetaMessage;
  tokens: LensTokenMessage[];
}

/**
 * Sensory / workspace / motor boundaries for one model.
 *
 * THERE IS DELIBERATELY NO DEFAULT VALUE AND NO CONSTANT ANYWHERE IN THIS
 * FEATURE. The reference implementation's L40/L90 are the source paper's
 * Sonnet-4.5 figures; BR-002 forbids porting them to another model and requires
 * the product make porting impossible by construction. Bands render only from a
 * report computed for the selected model, and are absent otherwise — a
 * fallback object would be those figures under another name.
 */
export interface BandReport {
  model: string;
  workspace_start: number;
  motor_start: number;
  /** How the boundaries were derived, surfaced next to the shading. */
  derivation: string;
}

/** Provenance behind the current readout (BR-007). */
export interface ReadoutProvenance {
  /** Absent for the logit lens, which involves no artifact at all. */
  artifact_id: string | null;
  target_layer?: string;
  attention_gradients?: string;
  target_position_scope?: string;
  aggregation?: string;
  corpus?: string;
  n_prompts?: number;
  seq_len?: number;
  dtype?: string;
}

/** One artifact as it exists on disk. PRESENCE, not validity. */
export interface JLensArtifactSummary {
  slug: string;
  directory: string;
  lens_file: string;
  size_bytes: number;
  has_config: boolean;
  /**
   * Layers this artifact covers.
   *
   * EMPTY MEANS UNKNOWN, not "none". An artifact whose config could not be
   * read still holds whatever it holds, and drawing that as zero coverage
   * would assert something the listing never checked.
   */
  layers?: number[];
  /**
   * Layers where J is the identity — the lens there IS the logit lens.
   *
   * The last decoder layer has no blocks after it, so its sub-network is the
   * identity by construction. A Diff at such a layer is empty because the two
   * lenses are the same lens, not because they happen to agree, and an empty
   * top row read without that context looks like a finding.
   */
  degenerate_layers?: number[];
  /**
   * Which block the Jacobian was taken TO.
   *
   * With a `penultimate` target a COMPLETE fit covers 0..N-2, so comparing
   * coverage against the model's layer count renders a full artifact as
   * "25/26" and colours it amber — reporting a recipe choice as a defect.
   */
  target_layer?: string | null;
}

export interface JLensCheckOutcome {
  check: string;
  /** 'pass' | 'fail' | 'not_run'. NOT_RUN is not a pass — see ValidationResponse. */
  status: string;
  detail: string;
  evidence?: Record<string, unknown>;
}

/**
 * The validation verdict, every class reported individually.
 *
 * `passed` is fail-closed and covers all six classes — the gate for handing an
 * artifact to an EXTERNAL consumer. Two of those classes need a live consumer
 * to run at all, so a validation performed from the workbench reports them
 * NOT_RUN and `passed` is false. That is the honest answer, not a defect, and
 * the UI says so rather than showing a red cross.
 */
export interface JLensValidationResponse {
  slug: string;
  passed: boolean;
  summary: string;
  results: JLensCheckOutcome[];
}

/**
 * Fixture for the SEMANTIC validation class.
 *
 * `expected_intermediate` must NOT appear in `prompt`: a token already present
 * is recovered by an artifact that encodes nothing at all, so a fixture that
 * breaks this passes against a broken lens. The server rejects it.
 */
export interface JLensSemanticProbe {
  prompt: string;
  expected_intermediate: string;
  /**
   * Omit to SCAN every fitted layer — the default.
   *
   * Naming a layer asserts WHERE an unspoken intermediate lives, which is a
   * property of the model and not something this project may assume (BR-002).
   * Two such defaults were shipped and both were wrong; the second discarded a
   * converged artifact whose mid-stack readout was the correct concept field.
   */
  layer?: number | null;
  /**
   * An unrelated prompt for which `expected_intermediate` would be absurd.
   *
   * A scan has more chances to hit than a single layer, so it needs a
   * false-positive control: if the token surfaces here too, the check FAILS
   * however well the real prompt scored.
   */
  control_prompt?: string | null;
  top_k?: number;
}

export interface JLensFitRequest {
  model_id: string;
  prompts: string[];
  layers?: number[] | null;
  freeze_qk?: boolean;
  corpus_name?: string;
  /** Without this NOTHING IS PUBLISHED — the suite fails closed on an unrun check. */
  semantic_probe?: JLensSemanticProbe | null;
  /**
   * Publish even when this fit is WEAKER evidence than the artifact it
   * replaces — fewer prompts, or not converged where the incumbent converged.
   *
   * Publishing is otherwise last-writer-wins, and "last" means finished last,
   * not best: a 400-prompt fit that never converged once published over a
   * 1097-prompt fit that did, because the weaker job had been queued hours
   * earlier and only got a worker once the queue drained.
   */
  allow_quality_regression?: boolean;
}

export interface JLensFitAccepted {
  task_id: string;
  model_id: string;
  queue: string;
}

/** A queued readout. The result arrives via the task, not this response. */
export interface ReadoutAccepted {
  task_id: string;
  model_id: string;
  status: string;
}

/**
 * A readout task's state.
 *
 * `readout` is null until `status` is SUCCESS. Treating a pending task as an
 * empty readout reproduces exactly the confusion this feature exists to
 * prevent, so `status` is always authoritative.
 */
export interface ReadoutResult {
  task_id: string;
  status: string;
  stage?: string | null;
  readout?: ReadoutResponse | null;
  error?: string | null;
}


/**
 * An intervention and the control it is meaningless without (BR-018).
 *
 * There is no way to request one WITHOUT a control: `k` and `control_seed`
 * define it and both are always sent. An intervention that moves the output
 * says nothing until compared with what a random direction of the same size
 * does.
 */
export interface JLensInterventionRequest {
  model_id: string;
  prompt: string;
  primitive: string;
  layers: number[];
  /** A SINGLE token; the server resolves its unembedding row. */
  direction_token?: string | null;
  /**
   * The token whose RANK is scored in the model's own output.
   *
   * Defaults server-side to `direction_token`. REQUIRED and must DIFFER for
   * `coordinate_swap`: a swap exchanges two coordinates, and one token would be
   * an additive steer wearing a swap's name — which is what the backend used to
   * run, labelled as a swap, before the hook grew a branch for it.
   */
  target_token?: string | null;
  /**
   * MORE PROMPTS, one TRIAL each.
   *
   * The result is a fraction of trials with a Wilson 95% interval; a single
   * prompt yields an interval spanning nearly the whole range, which is the
   * honest rendering of one observation and rarely a finding.
   */
  prompts?: string[] | null;
  strength?: number;
  k?: number;
  control_seed?: number;
  artifact_id?: string | null;
}

export interface JLensAnnotateRequest {
  /** All this needs. The rest is resolved server-side from the feature row. */
  feature_id: string;
  model_id?: string;
  sae_id?: string;
  layer?: number;
  label_tokens?: string[];
  top_k?: number;
}

export interface JLensAnnotation {
  feature_id: string;
  layer: number;
  lens_kurtosis: number | null;
  /** UNKNOWN without a band report for this model — a real answer, not a gap. */
  workspace_class: string;
  top_tokens: string[];
  disagreement_score?: number | null;
  has_disagreement?: boolean;
  evidence_rung?: number;
}

export interface JLensWatchlistRequest {
  name: string;
  artifact_ref: string;
  /** REQUIRED. A threshold applied to a differently computed score is a
   *  different detector, and the consumer cannot notice. */
  scoring_definition: string;
  concepts: Array<{ token: string; threshold: number }>;
  control_set?: string[];
}

export interface JLensWatchlistResponse {
  name: string;
  artifact_ref: string;
  scoring_definition: string;
  concept_count: number;
}

/** One hand-typed token, resolved against the model's own vocabulary. */
export interface JLensTokenCheck {
  token: string;
  /** The ids it encodes to — shown, because "[4874, 883]" is what makes
   *  "this is two tokens" concrete. */
  ids: number[];
  n_tokens: number;
  /** Usable as a lens direction, which is defined for exactly one id. */
  usable: boolean;
  detail: string;
}

/** One downloadable candidate in a HuggingFace repo. */
export interface JLensAcquireCandidate {
  path: string;
  size_bytes: number | null;
  /**
   * A `config.yaml` sits beside this file.
   *
   * PRESENCE, NOT A VERDICT. `check_weight_identity` still returns UNVERIFIED
   * when that config names no model, and REFUSES outright when it names a
   * different one — so this says identity CAN be looked for, not that it will
   * check out. The badge said "identity checkable" until a review pointed out
   * the same overclaim.
   */
  has_config: boolean;
  has_convergence: boolean;
  /**
   * Null when NO VERDICT WAS COMPUTED — which has three causes, not one.
   *
   * The endpoint computes one only `if dims and c.size_bytes`: no model named,
   * OR the Hub reported no size, OR the model row lacks the dimensions to
   * derive a bound. Treating null as "fine" permits exactly the download the
   * preview exists to prevent.
   */
  fits_envelope: boolean | null;
  envelope_detail: string | null;
}

export interface JLensAcquirePreview {
  repo_id: string;
  /** The RESOLVED commit. `main` moves, so an acquisition pinned to it is not a
   *  reproducible statement. */
  revision: string;
  candidates: JLensAcquireCandidate[];
}

export interface JLensAcquirePreviewRequest {
  repo_id: string;
  revision?: string;
  model_id?: string;
  access_token?: string;
}

export interface JLensAcquireRequest {
  model_id: string;
  repo_id: string;
  path_in_repo: string;
  revision?: string;
  access_token?: string;
  allow_coverage_loss?: boolean;
  allow_quality_regression?: boolean;
  replace_staged?: boolean;
}

export interface JLensPublishRequest {
  model_id: string;
  target_repo: string;
  access_token?: string;
  dataset?: string;
  create_repo?: boolean;
  private?: boolean;
}

File: backend/src/schemas/jlens.py
"""
Pydantic schemas for J-space lens readouts.

WIRE FORMAT IS NOT OURS TO DESIGN (BR-029, PADR IDL-45). These shapes mirror
Neuronpedia's lens stream exactly, so a miStudio stream and a Neuronpedia stream
are interchangeable at the client and the readout panel is driven by either with
no adaptation layer. Adopting the upstream shape also removes a whole class of
contract invention and satisfies part of the projection obligation structurally.

    meta  = { model, types, layers_by_type, top_n, prompt_len }
    token = { position, token, id, is_generated, results: slice[] }
    slice = { type, top_tokens[layer][k], top_probs[layer][k] }

Two details are load-bearing and are enforced here rather than by convention:

  * `top_tokens` entries are DECODED STRINGS, not token ids. Emitting ids
    type-checks fine against a looser schema and renders as unreadable cells in
    a client expecting text.
  * `layers_by_type` is per lens type and drives the client's layer axis. The
    reference panel hardcodes 21 layers at 0,5,...,100; the reference model has
    16. Nothing may assume a layer count or spacing.
"""

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

# The lens types the stream can carry. DIFF is a client-side rendering mode over
# two slices, never a transported type — the server does not emit it.
LensType = Literal["JACOBIAN_LENS", "LOGIT_LENS"]

# Readout mode (BR-008). These answer different questions and are not
# interchangeable; see LayerApplicability and DecompositionResult.
ReadoutMode = Literal["full_ranked", "probe", "decomposition"]


class LayerApplicability(BaseModel):
    """Which computations are meaningful at one layer.

    Layer kind is PER-LAYER state, not a model property. The reference model
    interleaves 10 convolutional layers with 6 attention layers, so
    "freeze Q/K" is undefined on 10 of 16 and attention-broadcast metrics are
    computable on 6.

    Inapplicable is expressed as None (absent), never False/0. A False would be
    averaged by downstream consumers and would silently understate; an absent
    value forces the consumer to decide what to do about it (BR-032).
    """

    model_config = ConfigDict(from_attributes=True)

    layer: int = Field(..., ge=0, description="Layer index within the model")
    has_attention: bool = Field(
        ..., description="Whether this layer performs attention at all"
    )
    frozen_qk_applicable: Optional[bool] = Field(
        None,
        description=(
            "Whether the frozen-Q/K recipe variant is meaningful here. None "
            "means INAPPLICABLE (e.g. a convolutional layer), which is not the "
            "same as False"
        ),
    )
    broadcast_metrics_applicable: Optional[bool] = Field(
        None,
        description=(
            "Whether attention-broadcast metrics can be computed here. None "
            "means inapplicable, not zero"
        ),
    )


class LensTypeSlice(BaseModel):
    """One lens type's readout for one token position, across layers.

    `top_tokens[layer_idx][k]` and `top_probs[layer_idx][k]` are parallel: the
    outer index is position in `meta.layers_by_type[type]`, NOT the model's
    absolute layer number.
    """

    model_config = ConfigDict(from_attributes=True)

    type: LensType
    top_tokens: List[List[str]] = Field(
        ...,
        description="Decoded token strings, [layer][k]. NOT token ids.",
    )
    top_probs: List[List[float]] = Field(
        ..., description="Probabilities parallel to top_tokens, [layer][k]"
    )

    @field_validator("top_tokens", mode="before")
    @classmethod
    def _tokens_must_be_strings(cls, v: Any) -> Any:
        """Reject token ids with an explanation, before pydantic's type error.

        mode="before" is deliberate. Pydantic already refuses int -> str, so an
        "after" validator here could NEVER FIRE — dead code that reads as
        protection. Running before type coercion means this fires first and
        says why the value is wrong, instead of "Input should be a valid
        string" against a bare integer.
        """
        if not isinstance(v, list):
            return v
        for layer_idx, row in enumerate(v):
            if not isinstance(row, list):
                continue
            for k, tok in enumerate(row):
                if not isinstance(tok, str):
                    raise ValueError(
                        f"top_tokens[{layer_idx}][{k}] is {type(tok).__name__} "
                        f"({tok!r}), expected a DECODED STRING. Emitting token "
                        "ids here type-checks against a looser schema and "
                        "renders as unreadable cells in the client."
                    )
        return v

    @model_validator(mode="after")
    def _shapes_must_agree(self) -> "LensTypeSlice":
        if len(self.top_tokens) != len(self.top_probs):
            raise ValueError(
                f"top_tokens has {len(self.top_tokens)} layers but top_probs has "
                f"{len(self.top_probs)}"
            )
        for i, (toks, probs) in enumerate(zip(self.top_tokens, self.top_probs)):
            if len(toks) != len(probs):
                raise ValueError(
                    f"layer index {i}: {len(toks)} tokens vs {len(probs)} probs"
                )
        return self


class LensMetaMessage(BaseModel):
    """Opening message of a readout stream."""

    model_config = ConfigDict(from_attributes=True)

    kind: Literal["meta"] = "meta"
    model: str = Field(..., description="Model identity the readout was taken from")
    types: List[LensType] = Field(..., min_length=1)
    layers_by_type: Dict[str, List[int]] = Field(
        ...,
        description=(
            "Absolute layer indices carried per lens type. Drives the client's "
            "layer axis; never assume a count or spacing"
        ),
    )
    top_n: int = Field(..., gt=0)
    prompt_len: int = Field(..., ge=0)

    # Per-layer applicability travels with the stream so a consumer never has to
    # infer homogeneity from the layer count (BR-032).
    layer_applicability: Optional[List[LayerApplicability]] = None

    @model_validator(mode="after")
    def _types_have_layers(self) -> "LensMetaMessage":
        missing = [t for t in self.types if t not in self.layers_by_type]
        if missing:
            raise ValueError(f"layers_by_type missing entries for {missing}")
        return self


class LensTokenMessage(BaseModel):
    """One token position, carrying one slice per lens type."""

    model_config = ConfigDict(from_attributes=True)

    kind: Literal["token"] = "token"
    position: int = Field(..., ge=0)
    token: str
    id: int
    is_generated: bool = False
    results: List[LensTypeSlice] = Field(..., min_length=1)


class LensDoneMessage(BaseModel):
    kind: Literal["done"] = "done"


class LensErrorMessage(BaseModel):
    """Terminal error. `reason` is required — a bare error message is unusable."""

    kind: Literal["error"] = "error"
    reason: str = Field(..., min_length=1)


class ReadoutRequest(BaseModel):
    """Request for a position x layer readout."""

    model_config = ConfigDict(from_attributes=True)

    model_id: str
    # Bounded deliberately. Readout cost is O(positions x layers x top_n) and
    # every position holds a d_model residual, so an unbounded prompt is a
    # memory-and-compute amplification from a single request.
    prompt: str = Field(..., min_length=1, max_length=8000)
    types: List[LensType] = Field(
        default_factory=lambda: ["LOGIT_LENS"], min_length=1, max_length=2
    )
    layers: Optional[List[int]] = Field(
        None,
        max_length=512,
        description="Absolute layer indices; None means every layer of the model",
    )
    top_n: int = Field(8, gt=0, le=100)

    # Identity of the J-lens artifact to use when types include JACOBIAN_LENS.
    # Absent is valid and means logit-only, which needs no artifact (BR-005).
    artifact_id: Optional[str] = None

    @model_validator(mode="after")
    def _jacobian_needs_an_artifact(self) -> "ReadoutRequest":
        if "JACOBIAN_LENS" in self.types and not self.artifact_id:
            raise ValueError(
                "JACOBIAN_LENS requested without artifact_id. The logit lens "
                "needs no artifact; the Jacobian lens does, and silently "
                "serving logit data under a Jacobian label is prohibited "
                "(BR-019 rung discipline)."
            )
        return self


class DecompositionProvenance(BaseModel):
    """Mandatory provenance for a sparse decomposition (BR-009).

    The decomposition is non-unique by construction: reproducibility is a
    provenance property, not a mathematical one. A figure without
    `control_seed` is INVALID rather than merely undocumented, because
    occupancy and excess-FVE are defined against a random-direction control.
    """

    model_config = ConfigDict(from_attributes=True)

    k: int = Field(..., gt=0, description="Sparsity level")
    solver: str = Field(..., min_length=1)
    solver_params: Dict[str, Any] = Field(default_factory=dict)
    iterations: int = Field(..., ge=0)
    convergence_criterion: str = Field(..., min_length=1)
    control_seed: int = Field(
        ...,
        description=(
            "Seed for the size-matched random-direction control. REQUIRED — "
            "occupancy and excess-FVE are defined as an excess over this "
            "control, so a figure without it cannot be reproduced or believed"
        ),
    )
    control_construction: str = Field(..., min_length=1)


class DecompositionResult(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    layer: int = Field(..., ge=0)
    position: int = Field(..., ge=0)
    active_tokens: List[str]
    coefficients: List[float]
    residual_norm: float
    provenance: DecompositionProvenance

    @model_validator(mode="after")
    def _active_set_is_parallel(self) -> "DecompositionResult":
        if len(self.active_tokens) != len(self.coefficients):
            raise ValueError(
                f"{len(self.active_tokens)} active tokens vs "
                f"{len(self.coefficients)} coefficients"
            )
        if len(self.active_tokens) > self.provenance.k:
            raise ValueError(
                f"active set of {len(self.active_tokens)} exceeds sparsity "
                f"level k={self.provenance.k}"
            )
        return self


class ProbeRequest(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    model_id: str
    prompt: str = Field(..., min_length=1, max_length=8000)
    tokens: List[str] = Field(
        ...,
        min_length=1,
        max_length=256,
        description="Named directions to score",
    )
    layers: Optional[List[int]] = Field(None, max_length=512)
    artifact_id: Optional[str] = None


class ProbeScore(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    layer: int
    position: int
    token: str
    score: float


class JLensArtifactRecipe(BaseModel):
    """Construction recipe, sufficient to rebuild the artifact (BR-007)."""

    model_config = ConfigDict(from_attributes=True)

    # DEFAULTS MATCH THE PAPER, and the fitter now implements it. An earlier
    # revision "fixed" these the wrong way round: the fitter took one source
    # position on a length-1 sub-network, so the schema was edited to say
    # `self_only_isolated` and `final` to match the CODE. That made schema and
    # implementation agree while both disagreed with the source. Alignment is
    # against the paper; the code moved to meet it.
    target_layer: Literal["final", "penultimate"] = "penultimate"
    attention_gradients: Literal["full", "frozen_qk"] = "full"
    target_position_scope: Literal[
        "self_only", "self_only_isolated", "future_only", "all_subsequent"
    ] = "all_subsequent"
    aggregation: Literal["mean", "median"] = "mean"
    corpus: str
    n_prompts: int = Field(..., ge=100, description="Floor of 100 (Appendix A.2)")
    seq_len: int = Field(..., gt=0)
    convergence_delta: Optional[float] = None
    dtype: Literal["fp16"] = "fp16"
    library_versions: Dict[str, str] = Field(default_factory=dict)

    # Recorded per layer, because a recipe choice can be inapplicable to a layer
    # rather than merely unset (BR-032). An artifact must not be described as
    # "frozen-Q/K" wholesale when the treatment reached only a subset.
    per_layer_applicability: List[LayerApplicability] = Field(default_factory=list)


class JLensArtifact(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    artifact_id: str
    model_id: str
    d_model: int = Field(..., gt=0)
    n_layers: int = Field(..., gt=0)
    n_vocab: int = Field(..., gt=0)
    layers: List[int]
    size_bytes: int = Field(..., gt=0)
    recipe: JLensArtifactRecipe

    def expected_size_bytes(self, dtype_bytes: int = 2) -> int:
        """Envelope bound derived from THIS model's dimensions.

        Never a constant: the required-vs-materialised ratio scales with
        vocabulary, so a bound hardcoded for one model passes on another while
        missing a real materialisation. Reference model ~32x, a 256k-vocab
        model ~111x (PADR IDL-42).
        """
        return self.d_model * self.d_model * dtype_bytes * self.n_layers

    def materialized_size_bytes(self, dtype_bytes: int = 2) -> int:
        """What the PROHIBITED materialised dictionary would cost."""
        return self.n_vocab * self.d_model * dtype_bytes * self.n_layers
